tests/workflow/test_run.py

```diff
@@ -0,0 +1,957 @@
+            DefaultExecutionConfig,
+            gradient_method=qml.gradients.param_shift,
+        ),
+        Shots((100000, 100000)),
+    ],
+    # 7
+    [
+        "reference.qubit",
+        replace(
+            DefaultExecutionConfig,
+            gradient_method=qml.gradients.param_shift,
+        ),
+        Shots(100000),
+    ],
+    # 8
+    [
+        "reference.qubit",
+        replace(
+            DefaultExecutionConfig,
+            gradient_method=qml.gradients.param_shift,
+        ),
+        Shots(None),
+    ],
+    # 9
+    [
+        "param_shift.qubit",
+        replace(
+            DefaultExecutionConfig,
+            gradient_method="device",
+            use_device_jacobian_product=False,
+        ),
+        Shots((100000, 100000)),
+    ],
+    # 10
+    [
+        "param_shift.qubit",
+        replace(
+            DefaultExecutionConfig,
+            gradient_method="device",
+            use_device_jacobian_product=False,
+        ),
+        Shots(100000),
+    ],
+    # 11
+    [
+        "param_shift.qubit",
+        replace(
+            DefaultExecutionConfig,
+            gradient_method="device",
+            use_device_jacobian_product=False,
+        ),
+        Shots(None),
+    ],
+]
+
+
+class TestNoInterfaceRequired:
+
+    def test_numpy_interface(self, seed):
+        """Test that tapes are executed correctly with the NumPy interface."""
+        container = TransformContainer(merge_rotations)
+        inner_tp = TransformProgram((container,))
+        device = qml.device("default.qubit", seed=seed)
+        tapes = [
+            QuantumScript(
+                [qml.RX(pnp.pi, wires=0), qml.RX(pnp.pi, wires=0)], [qml.expval(qml.PauliZ(0))]
+            )
+        ]
+        config = ExecutionConfig(interface="numpy", gradient_method=qml.gradients.param_shift)
+        results = run(tapes, device, config, inner_tp)
+
+        assert qml.math.get_interface(results) == "numpy"
+        assert qml.math.allclose(results[0], 1.0)
+
+    @pytest.mark.torch
+    @pytest.mark.parametrize(
+        "interface, gradient_method",
+        [("torch", None), ("torch", "backprop")],
+    )
+    def test_no_gradient_computation_required(self, interface, gradient_method, seed):
+        """Test that tapes execute without an ML boundary when no gradient computation is required."""
+        container = TransformContainer(merge_rotations)
+        inner_tp = TransformProgram((container,))
+        device = qml.device("default.qubit", seed=seed)
+        tapes = [
+            QuantumScript(
+                [qml.RX(pnp.pi, wires=0), qml.RX(pnp.pi, wires=0)], [qml.expval(qml.PauliZ(0))]
+            )
+        ]
+        config = ExecutionConfig(interface=interface, gradient_method=gradient_method)
+        results = run(tapes, device, config, inner_tp)
+
+        assert qml.math.get_interface(results) == "numpy"
+        assert qml.math.allclose(results[0], 1.0)
+
+
+torch = pytest.importorskip("torch")
```

Comment:
I'm pretty sure this would apply to the whole file, so this file won't run in the CI at all.

Instead, we should probably just `import torch` inside each individual test case.

Reply:
Right ... once I move the tests into their own files it'll be as intended 😄 .